openstack/datadog_checks/openstack/openstack.py

```diff
@@ -844,24 +856,53 @@ def get_stats_for_single_hypervisor(self, hyp_id, instance, host_tags=None):
             for i, avg in enumerate([1, 5, 15]):
                 self.gauge('openstack.nova.hypervisor_load.{0}'.format(avg), load_averages[i], tags=tags)
 
-    def get_all_server_ids(self, filter_by_host=None):
+    # Get all of the server IDs and their metadata and cache them
+    # After the first run, we will only get servers that have changed state since the last collection run
+    def get_all_servers(self, i_key, filter_by_host=None):
         query_params = {}
         if filter_by_host:
             query_params["host"] = filter_by_host
 
-        url = '{0}/servers'.format(self.get_nova_endpoint())
+        # If we don't have a timestamp for this instance, default to None
+        if i_key not in self.changes_since_time:
+            self._changes_since_time[i_key] = None
+
+        if self.changes_since_time[i_key]:
+            query_params['changes-since'] = changes_since_time
+
+        # Since we ultimately only perform actions on ACTIVE servers
+        # Lets filter to only those to reduce the payload we receive
+        active_query_params['status'] = 'ACTIVE,DELETED,SHUTOFF'
```

Comment:
```python
','.join(DIAGNOSTICABLE_STATES+REMOVED_STATES)
```

Reply:
It turns out that this isn't available for all versions of the API, so I'm making 2 addition API calls per check run to get the SHUTOFF and DELETED VMs. This shouldn't be an issue because its not called per server, so it shouldn't be a large overhead.